// tailwindcss-theming/dist/Theming/Color/Color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tinycolor_1 = require("@ctrl/tinycolor");
class Color {
    constructor(name, value) {
        this.name(name || 'transparent');
        this.value(value || 'transparent');
    }
    /**
     * Gets the color name.
     *
     * @readonly
     * @type {string}
     * @memberof Color
     */
    get keyName() {
        return this._name;
    }
    /**
     * Sets the name of this color.
     * @param name Color name.
     *
     * @returns self
     */
    name(name) {
        this._name = name;
        return this;
    }
    /**
     * Sets the value of this color.
     * @param name Color name.
     *
     * @returns self
     */
    value(value) {
        this._value = value;
        this._computed = new tinycolor_1.TinyColor(value);
        return this;
    }
    /**
     * Gets the color.
     *
     * @return TinyColor
     */
    get color() {
        return this._computed;
    }
    /**
     * Gets the computed color.
     *
     * @readonly
     * @memberof Color
     */
    get computed() {
        return this._computed;
    }
    /**
     * Gets the original value.
     *
     * @readonly
     * @memberof Color
     */
    get original() {
        return this._value;
    }
}
exports.Color = Color;
